### PR TITLE
dubbo_proxy: Add hessian rpc deserialize

### PR DESCRIPTION
**Pull request by zyfjeff**
Signed-off-by: tianqian.zyf <tianqian.zyf@alibaba-inc.com>

*Description*: Add hessian rpc deserialize
*Risk Level*: low
*Testing*: unittests
*Docs Changes*: N/A
*Release Notes*: N/A 
[Optional Fixes #Issue]  #3998  #4200

### CODE REVIEW

File: source/extensions/filters/network/dubbo_proxy/hessian_serializer_impl.cc [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+enum class RpcResponseType : unsigned char {
+  ResponseWithException = 0,
+  ResponseWithValue = 1,
+  ResponseWithNullValue = 2,
+  ResponseWithExceptionWithAttachments = 3,
+  ResponseValueWithAttachments = 4,
+  ResponseNullValueWithAttachments = 5,
+
+  // ATTENTION: MAKE SURE THIS REMAINS EQUAL TO THE LAST RpcResponseType TYPE
+  LastResponseType = ResponseNullValueWithAttachments,
```
Comment (lizan) [inline source/extensions/filters/network/dubbo_proxy/hessian_serializer_impl.cc]: where is this used?

File: source/extensions/filters/network/dubbo_proxy/hessian_serializer_impl.cc [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+#include "extensions/filters/network/dubbo_proxy/hessian_utils.h"
+#include "extensions/filters/network/dubbo_proxy/serialization.h"
+#include "extensions/filters/network/dubbo_proxy/serialization_impl.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace DubboProxy {
+
+enum class RpcResponseType : unsigned char {
```
Comment (lizan) [inline source/extensions/filters/network/dubbo_proxy/hessian_serializer_impl.cc]: : uint8_t?

File: source/extensions/filters/network/dubbo_proxy/serialization.h [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+namespace NetworkFilters {
+namespace DubboProxy {
+
+/**
+ * Names of available serialization implementations.
+ */
+class SerializerNameValues {
+public:
+  // hessian serialization
+  const std::string HESSIAN = "hessian";
```
Comment (lizan) [inline source/extensions/filters/network/dubbo_proxy/serialization.h]: nit: prefer CamelCase for short constant

File: source/extensions/filters/network/dubbo_proxy/serialization.h [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+typedef std::unique_ptr<RpcResult> RpcResultPtr;
+
+class SerializationCallbacks {
+public:
+  virtual ~SerializationCallbacks() {}
+  virtual void onRpcInvocation(RpcInvocationPtr&& invo) PURE;
+  virtual void onRpcResult(RpcResultPtr&& res) PURE;
+};
+
+class Serialization {
```
Comment (lizan) [inline source/extensions/filters/network/dubbo_proxy/serialization.h]: (throughout this PR) this should be named `Deserializer`?

File: source/extensions/filters/network/dubbo_proxy/serialization.h [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+  virtual const std::string& name() const PURE;
+  /**
+   * deserialize an rpc call, potentially invoking callbacks
+   * If successful, the RpcInvocation removed from the buffer
+   *
+   * @param buffer the currently buffered dubbo data
+   * @body_size the complete RpcInvocation size
+   * @throws EnvoyException if the data is not valid for this serialization
+   */
+  virtual void deserializeRpcInvocation(Buffer::Instance& buffer, size_t body_size) PURE;
```
Comment (lizan) [inline source/extensions/filters/network/dubbo_proxy/serialization.h]: I don't know why you need a callback. Can this just return `RpcInvocationPtr`? Looked at #4200 too but I don't think there is a gain from this structure.

Comment (zyfjeff) [reply source/extensions/filters/network/dubbo_proxy/serialization.h]: The purpose of using callback here is to decouple the decoder and filter. All stats statistics are implemented in the filter. The decoder is only responsible for protocol parsing, and then the callback will parse the parsed result to the filter.

Comment (zyfjeff) [reply source/extensions/filters/network/dubbo_proxy/serialization.h]: The purpose of using callback here is to decouple the decoder and filter. All stats statistics are implemented in the filter. The decoder is only responsible for protocol parsing, and then the callback will parse the parsed result to the filter.

Comment (zyfjeff) [reply source/extensions/filters/network/dubbo_proxy/serialization.h]: The purpose of using callback here is to decouple the decoder and filter. All stats statistics are implemented in the filter. The decoder is only responsible for protocol parsing, and then the callback will parse the parsed result to the filter.

Comment (zyfjeff) [reply source/extensions/filters/network/dubbo_proxy/serialization.h]: The purpose of using callback here is to decouple the decoder and filter. All stats statistics are implemented in the filter. The decoder is only responsible for protocol parsing, and then the callback will parse the parsed result to the filter.

Comment (zyfjeff) [reply source/extensions/filters/network/dubbo_proxy/serialization.h]: The purpose of using callback here is to decouple the decoder and filter. All stats statistics are implemented in the filter. The decoder is only responsible for protocol parsing, and then the callback will parse the parsed result to the filter.

Comment (zyfjeff) [reply source/extensions/filters/network/dubbo_proxy/serialization.h]: The purpose of using callback here is to decouple the decoder and filter. All stats statistics are implemented in the filter. The decoder is only responsible for protocol parsing, and then the callback will parse the parsed result to the filter.

Comment (zyfjeff) [reply source/extensions/filters/network/dubbo_proxy/serialization.h]: The purpose of using callback here is to decouple the decoder and filter. All stats statistics are implemented in the filter. The decoder is only responsible for protocol parsing, and then the callback will parse the parsed result to the filter.

Comment (zyfjeff) [reply source/extensions/filters/network/dubbo_proxy/serialization.h]: The purpose of using callback here is to decouple the decoder and filter. All stats statistics are implemented in the filter. The decoder is only responsible for protocol parsing, and then the callback will parse the parsed result to the filter.

Comment (zyfjeff) [reply source/extensions/filters/network/dubbo_proxy/serialization.h]: The purpose of using callback here is to decouple the decoder and filter. All stats statistics are implemented in the filter. The decoder is only responsible for protocol parsing, and then the callback will parse the parsed result to the filter.

File: source/extensions/filters/network/dubbo_proxy/serialization.h (hunk 4 above)
Comment (zyfjeff) [reply source/extensions/filters/network/dubbo_proxy/serialization.h]: fixed

File: source/extensions/filters/network/dubbo_proxy/serialization.h (hunk 3 above)
Comment (zyfjeff) [reply source/extensions/filters/network/dubbo_proxy/serialization.h]: fixed

File: source/extensions/filters/network/dubbo_proxy/hessian_serializer_impl.cc (hunk 2 above)
Comment (zyfjeff) [reply source/extensions/filters/network/dubbo_proxy/hessian_serializer_impl.cc]: fixed

**Comment by zyfjeff** [comment]
@lizan  Sorry, there was a problem with github yesterday, so it was not updated in time.

File: source/extensions/filters/network/dubbo_proxy/deserializer.h [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+ */
+class RpcResult {
+public:
+  virtual ~RpcResult() {}
+  virtual bool hasException() const PURE;
+};
+
+typedef std::unique_ptr<RpcResult> RpcResultPtr;
+
+class DeserializationCallbacks {
```
[the remark this answers is not in the document]
Comment (zyfjeff) [reply source/extensions/filters/network/dubbo_proxy/deserializer.h]: What you mean is to let the `Deserializer` return `RpcInvocation/RpcResult` directly, and call `DecodeCallback` in the `Decoder` to call `onRpcInvocation/onRpcResult` ?

File: source/extensions/filters/network/dubbo_proxy/hessian_deserializer_impl.cc [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+  size_t total_size = 0;
+  bool has_value = true;
+
+  RpcResultPtr result;
+  RpcResponseType type = static_cast<RpcResponseType>(HessianUtils::peekInt(buffer, &total_size));
+
+  switch (type) {
+  case RpcResponseType::ResponseWithException:
+  case RpcResponseType::ResponseWithExceptionWithAttachments:
+    result = std::make_unique<RpcResultImpl>(true);
```
[the remark this answers is not in the document]
Comment (zyfjeff) [reply source/extensions/filters/network/dubbo_proxy/hessian_deserializer_impl.cc]: fixed

**Comment by billowqiu** [comment]
@zyfjeff dubbo_proxy hessian rpc deserialize is not implement the full dubbo protocol？
https://github.com/apache/incubator-dubbo/blob/master/dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcInvocation.java

```
    private Class<?>[] parameterTypes;
    private Object[] arguments;
    private Map<String, String> attachments;
```
these fileds has not deserialized.

**Comment by zyfjeff** [comment]
@billowqiu 
Yes, the current implementation of the community's hessian2 only deserializes the basic types. It is not supported for the Object type. Our company has already done it internally and is slowly submitting it to the community.

**Comment by billowqiu** [comment]
I am investigating the c++ call dubbo service，hopeful see your ali internally implementation.
have any plan to  submit it？
顺便能加个wechat不，^_^。

**Comment by zyfjeff** [comment]
@billowqiu 
We are submitting to the community,  my wechat 19941246896

**Comment by abel123** [comment]
> @billowqiu
> Yes, the current implementation of the community's hessian2 only deserializes the basic types. It is not supported for the Object type. Our company has already done it internally and is slowly submitting it to the community.

can I see the related public code anywhere?

**Comment by abel123** [comment]
@zyfjeff

**Comment by zyfjeff** [comment]
@abel123 I'm sorry, there is no open source code, once open source I will ping you

**Comment by serverglen** [comment]
> I'm sorry, there is no open source code, once open source I will ping you

@billowqiu Is it open source now?

**Comment by serverglen** [comment]
> @billowqiu
> We are submitting to the community, my wechat 19941246896

@billowqiu Is it submitted to the community now?

**Pull request merged**